src/core/cgroup.c

```diff
@@ -3110,8 +3110,8 @@ static int on_cgroup_empty_event(sd_event_source *s, void *userdata) {
         if (!u)
                 return 0;
 
-        assert(u->in_cgroup_empty_queue);
-        u->in_cgroup_empty_queue = false;
+        assert(u->in_cgroup_empty_queue || !strv_isempty(u->empty_cgroups));
+
         LIST_REMOVE(cgroup_empty_queue, m->cgroup_empty_queue, u);
```

Comment:
this is not allowed to be called if the unit is not actually in the queue you remove it from here.

Reply:
Actually, I checked my code again and the `assert` just above ensures that the unit is in the `cgroup_empty_queue`.

When I add elements in the `u->empty_cgroups` vector, I put the unit `u` in the `cgroup_empty_queue`, but I don't set `u->in_cgroup_empty_queue` to `true`.

In that way, we can distinguish two cases:
- the unit is in the `cgroup_empty_queue` because its cgroup is empty (`u->in_cgroup_empty_queue` is `true`)
- the unit is in the `cgroup_empty_queue` because it was delegated cgroups (`u->empty_cgroups` is not empty)